### PR TITLE
Cell as instance of Arrow

### PR DESCRIPTION
**Pull request by ostera**
#### Original Intent
I'm hoping this PR will end up with a good implementation of Cell as an instance of Arrow (similar to how different instances of Functor reimplement their necessary methods) in order to provide elegant compositional capabilities.

## Summary

This PR as is includes a very simple instance of a Cell as a composition of arrows of the form:

`( id (+++) (( f (***) id ) >>> retag) ) >>> flatten`.

In reality a Cell becomes a function from `Either (a,b) (a,b) -> Either (a,b) (a,b)`, where `Left` represents a _same in, same out_ result, whereas a `Right` represents a different expected output.

As an example, calling a Cell with `Left.of([1,2])` will bypass the arrow, and calling it with `Right.of([1,2])` will run the cell's function with the first value, and compare the output with the second value and return either a `Left.of([in,out])` for unchanged output or a `Right.of([in,out])` for a different output.

This makes it trivial to build dependent computations that should stop at the point where one of the cells returns a result it already has returned before (which explains the bypassing in the function definition).

### Computation Graph

As part of this definition, we can now explore the development of a simple computational graph that will only recompute values whenever their inputs have changed (inspired by Jane Street's Incrementals OCaml library).

## Left To Do

- [x] Tests for Cell
- [x] runCell function
- [x] Tests for ~~Node~~ Graph
- [x] Calling the Graph recomputes a subset of it
- [x] ~~Introduce a Node type~~
- [x] ~~runGraph function to compute a graph~~
- [x] ~~runGraph may take a single value and build up the rest of the inputs~~~

### CODE REVIEW

File: src/cell.js [hunk 1]
```diff
@@ -1,44 +1,83 @@
...
+    return cond(
+      [() => fx === h, Left([h, t])],
+      [true, Right([fx, t])])
   }
 
-  const _dirty = args => () => cell_fn.last_args[0] !== args [0]
+  // first comes from product
+  // second comes from first
+
+  _c.compose = g => _c.compose(g.sum(Cell(id)))
```
Comment (ostera) [inline src/cell.js]: Wait, why is `_c.compose` depending on `_c.compose` ?

Comment (xaviervia) [inline src/cell.js]: Lol. Got distracted. it doesn’t have to.

**Comment by xaviervia** [comment]
```javascript
import {
  Arrow,
} from './arrow'

import {
  Left,
  Right,
  either,
} from './either'

import {
  match,
} from './cond'

import {
  type,
} from './type'

const id = x => x

const Stable = type('Stable')
const Recompute = type('Recompute')

const Cell = f => {
  const c = Arrow(
    match({
      Stable: x => Arrow(id).product(Arrow(id).pipe(Stable.of))(x)[1],
      Recompute: ([arg, [expected, tail]]) => (
        result => result === expected
          ? Stable.of([expected, tail])
          : Recompute.of([result, tail])
      )(f(arg))
    })
  )

  c.pipe = g =>
    x => match({
      Stable: y => Stable.of(y),
      Recompute: ([head, [expected, tail]]) => Recompute.of([
        head,
        g(Recompute.of([head, tail]))
      ])
    })(c(x))

  return c
}

// Cell.of =

const a = Cell( x => `${x} world` )
const b = Cell( x => `${x} of Cells`)

const result = a.pipe(b)(
  Recompute.of([
    'hello',
    [
      'hello monica',
      [
        'hello world of Cells',
        [
          'a',
          null
        ]
      ]
    ]
  ])
)

console.log(result)
console.log(result['@@value'])
```

**Comment by xaviervia** [comment]
```javascript
import {
  Arrow,
} from './arrow'

import {
  Left,
  Right,
  either,
} from './either'

import {
  match,
} from './cond'

import {
  type,
} from './type'

const id = x => x

const Stable = type('Stable')
const Recompute = type('Recompute')

const Cell = (f, next) => Arrow(
    match({
      Stable: x => Arrow(id).product(Arrow(id).pipe(Stable.of))(x)[1],
      Recompute: ([arg, [expected, tail]]) => (
        result => result === expected
          ? Stable.of([expected, tail])
          : Recompute.of([
            result,
            next(Recompute.of([result, tail]))
          ])
      )(f(arg))
    })
  )

const c = Cell( x => x.toUpperCase(), x => x)
const b = Cell( x => `${x} of Cells`, c )
const a = Cell( x => `${x} world`, b )

const result = a(Recompute.of(['hello', ['not expected', ['wat', ['HELLO WORLD OF CELLS', []]]]]))

console.log(result)
console.log(result['@@value'])
console.log(result['@@value'][1]['@@value'])
console.log(result['@@value'][1]['@@value'][1]['@@value'])
```

File: test.js [hunk 2]
```diff
@@ -8,34 +8,56 @@ import {
...
+    Right.of([4, 56]),
+  ],
+  Right.of([4, 104]),
+  [
+    Right.of([4, 14]),
+    Right.of([4, 24]),
+    Right.of([4, 46]),
+    Right.of([4, 56]),
+  ],
+]))
```
Comment (ostera) [inline test.js]: Output: https://gist.github.com/ostera/b52a9dddf342f19982addb78e054c511

**Pull request merged**